src/clustering.jl

```diff
@@ -0,0 +1,351 @@
+            temp = (intersection_number / N) *
+                   log(N * intersection_number / (true_number * pred_number))
+            if !isnan(temp)
+                  result += temp
+            end
+        end
+    end
+    result
+end
+
+"""
+    normalized_mutual_info_score(target, output, mode)
+normalized_mutual_info_score computes mutual information but it limits the range of the
+return value from 0 (no similarity) to 1 (completely similar). There are
+five different approaches introduced by Vinh et al. (2010), that they can
+be selected by 'mode' argument.
+"""
+function normalized_mutual_info_score(target::AbstractVector,
+                                      output::AbstractVector,
+                                      mode::AbstractString = "sqrt")
+    result = 0
+    if mode == "sqrt"
+        denominator = sqrt(entropy(target) * entropy(output))
+        if denominator == 0
+            result = 0.0
+        else
+            result = mutual_info_score(target, output) /
+                     denominator
+        end
+    elseif mode == "max"
+        denominator = max(entropy(target), entropy(output))
+        if denominator == 0
+            result = 0.0
+        else
+            result = mutual_info_score(target, output) /
+                     denominator
+        end
+    elseif mode == "min"
+        denominator = min(entropy(target), entropy(output))
+        if denominator == 0
+            result = 0.0
+        else
+            result = mutual_info_score(target, output) /
+                     denominator
+        end
+    elseif mode == "sum"
+        denominator = entropy(target) + entropy(output)
+        if denominator == 0
+            result = 0.0
+        else
+            result = 2 * mutual_info_score(target, output) /
+                     denominator
+        end
+    elseif mode == "joint"
+        denominator = entropy(target, output)
+        if denominator == 0
+            result = 0.0
+        else
+            result = mutual_info_score(target, output) /
+                     denominator
+        end
+    else
+        throw(DomainError())
+    end
+    result
+end
+
+"""
+    expected_mutual_informatio(target, output)
+expected_mutual_informatio returns the expected value of mutual
+information which is defined in Vinh et al. (2010).
+"""
+function expected_mutual_information(U::AbstractVector,
+                                     V::AbstractVector)
+    result = 0
+    table = ContingencyTable(U, V)
+    N = table.N
+    a = table.a
+    b = table.b
+    for i = 1:table.R
+        for j = 1:table.C
+            for nij = max(1, a[i] + b[j] - N) : min(a[i], b[j])
+                result += (nij / N) * log( N*nij / (a[i] * b[j])) *
+                (f(a[i]) * f(b[j]) * f(N - a[i]) * f(N - b[j])) /
+                (f(N) * f(nij) * f(a[i] - nij) * f(b[j] - nij) *
+                f(N - a[i] - b[j] + nij))
+            end
+        end
+    end
+    result
+end
+
+f(x) = factorial(x)
```

Comment:
I'd rather not have this. I can see why you use it, but I don't think its worth it